app/upgrades/v4/upgrade.go

```diff
@@ -83,3 +59,132 @@ func CreateUpgradeHandler(
+			keyTable = slashingtypes.ParamKeyTable()
+		case govtypes.ModuleName:
+			keyTable = govv1.ParamKeyTable()
+		case crisistypes.ModuleName:
+			keyTable = crisistypes.ParamKeyTable()
+
+		// Ethermint  modules
+		case evmtypes.ModuleName:
+			keyTable = evmtypes.ParamKeyTable()
+		case feemarkettypes.ModuleName:
+			keyTable = feemarkettypes.ParamKeyTable()
+		default:
+			continue
+		}
+
+		if !subspace.HasKeyTable() {
+			subspace.WithKeyTable(keyTable)
+		}
+	}
+	// Migrate Tendermint consensus parameters from x/params module to a dedicated x/consensus module.
+	baseAppLegacySS := keepers.ParamsKeeper.Subspace(baseapp.Paramspace).WithKeyTable(paramstypes.ConsensusParamsKeyTable())
+	baseapp.MigrateParams(ctx, baseAppLegacySS, &keepers.ConsensusParamsKeeper)
+}
+
+func migrateDelayedAckParams(ctx sdk.Context, delayedAckKeeper delayedackkeeper.Keeper) {
+	// overwrite params for delayedack module due to added parameters
+	params := delayedacktypes.DefaultParams()
+	delayedAckKeeper.SetParams(ctx, params)
+}
+
+func migrateRollappParams(ctx sdk.Context, rollappkeeper *rollappkeeper.Keeper) {
+	// overwrite params for rollapp module due to proto change
+	params := rollapptypes.DefaultParams()
+	params.DisputePeriodInBlocks = rollappkeeper.DisputePeriodInBlocks(ctx)
+	rollappkeeper.SetParams(ctx, params)
+}
+
+func migrateRollapps(ctx sdk.Context, rollappkeeper *rollappkeeper.Keeper) error {
+	list := rollappkeeper.GetAllRollapps(ctx)
+	for _, oldRollapp := range list {
+		newRollapp := ConvertOldRollappToNew(oldRollapp)
+		if err := newRollapp.ValidateBasic(); err != nil {
+			return err
+		}
+		rollappkeeper.SetRollapp(ctx, newRollapp)
+	}
+	return nil
+}
+
+func migrateSequencers(ctx sdk.Context, sequencerkeeper sequencerkeeper.Keeper) {
+	list := sequencerkeeper.GetAllSequencers(ctx)
+	for _, oldSequencer := range list {
+		newSequencer := ConvertOldSequencerToNew(oldSequencer)
+		sequencerkeeper.SetSequencer(ctx, newSequencer)
+	}
+}
+
+func ConvertOldRollappToNew(oldRollapp rollapptypes.Rollapp) rollapptypes.Rollapp {
+	bech32Prefix := oldRollapp.RollappId[:5]
+	alias := strings.Split(oldRollapp.RollappId, "_")[0]
+	return rollapptypes.Rollapp{
+		RollappId:        oldRollapp.RollappId,
+		Creator:          oldRollapp.Creator,
+		GenesisState:     oldRollapp.GenesisState,
+		ChannelId:        oldRollapp.ChannelId,
+		Frozen:           oldRollapp.Frozen,
+		RegisteredDenoms: oldRollapp.RegisteredDenoms,
+		Bech32Prefix:     bech32Prefix,                                        // placeholder data
+		GenesisChecksum:  string(crypto.Sha256([]byte(oldRollapp.RollappId))), // placeholder data
+		Alias:            alias,                                               // placeholder data
+		Metadata: &rollapptypes.RollappMetadata{
+			Website:          "", // TODO
+			Description:      "", // TODO
+			LogoDataUri:      "", // TODO
+			TokenLogoDataUri: "", // TODO
+			Telegram:         "", // TODO
+			X:                "", // TODO
```

Comment:
?

Reply:
commented about this a bunch of times already

Reply:
ah sorry